clash-cores/src/Clash/Cores/Crc/Internal.hs

```diff
@@ -0,0 +1,688 @@
+      --   '_crcReflectInput' and '_crcReflectOutput' to perform a least
+      --   significant bit first computation.
+      , _crcInitial       :: BitVector crcWidth
+      -- ^ Initial value of CRC register at reset. Most significant bit always
+      --   corresponds to the highest order term in the CRC register.
+      , _crcReflectInput  :: Bool
+      -- ^ If True, the input data words are bit-reflected, so that they are
+      --   processed least significant bit first.
+      , _crcReflectOutput :: Bool
+      -- ^ If True, the output CRC is bit-reflected, so the least-significant bit
+      --   of the output is the highest-order bit of the CRC register.
+      --   Note that this reflection is performed over the entire CRC register;
+      --   for transmission you may want to treat the output as a little-endian
+      --   multi-word value, so for example the reflected 16-bit output 0x4E4C
+      --   would be transmitted as the two octets 0x4C 0x4E, each transmitted
+      --   least significant bit first.
+      , _crcXorOutput     :: BitVector crcWidth
+      -- ^ The output CRC will be the CRC register XOR'd with this value, applied
+      --   after any output bit-reflection.
+      }
+  deriving (Generic, Show, ShowX, Lift)
+
+-- | No domain in 'CrcParams'
+type instance TryDomain t (CrcParams crcWidth dataWidth) = 'NotFound
+
+-- | This class is used to define CRCs
+--
+-- First make a data declaration without constructors for your CRC and then
+-- create a 'KnownCrc' instance for that type which contains your CRC parameters:
+--
+-- @
+-- data MyCrc
+-- instance KnownCrc MyCrc where
+--  type CrcWidth MyCrc = ...
+--  crcParams _ dataWidth = CrcParams
+--    { _crcWidth = SNat
+--    , _crcDataWidth = dataWidth
+--    , _crcPolynomial = ..
+--    , _crcInitial = ..
+--    , _crcReflectInput = ..
+--    , _crcReflectOutput = ..
+--    , _crcXorOutput = ..
+--    }
+-- @
+--
+-- See "Clash.Cores.Crc.Catalog" for many definition examples.
+class (KnownNat (CrcWidth crc), 1 <= (CrcWidth crc)) => KnownCrc (crc :: Type) where
+  type CrcWidth crc :: Nat
+  crcParams :: Proxy crc -> SNat dataWidth -> CrcParams (CrcWidth crc) dataWidth
+
+-- | Allow one to compute values for the CRC
+--
+-- __NB__: NOT for use in hardware
+data SoftwareCrc (crcWidth :: Nat) (dataWidth :: Nat)
+  = SoftwareCrc
+      { _crcParams      :: CrcParams crcWidth dataWidth
+      , _crcTopBitMask  :: BitVector (crcWidth + dataWidth)
+      , _crcpolyShifted :: BitVector (crcWidth + dataWidth)
+      , _crcCurrent     :: BitVector (crcWidth + dataWidth)
+      }
+  deriving (Generic, Show, ShowX)
+
+-- | No domain in 'SoftwareCrc'
+type instance TryDomain t (SoftwareCrc crcWidth dataWidth) = 'NotFound
+
+-- | Apply function only when bool is True
+applyWhen :: Bool -> (a -> a) -> a -> a
+applyWhen True f x = f x
+applyWhen False _ x = x
+
+-- | Reverse bitvector
+reverseBV :: KnownNat n => BitVector n -> BitVector n
+reverseBV = v2bv . reverse . bv2v
```

Comment:
Once this PR is merged could you raise an issue to 
1. add this function to `clash-prelude` and
2. make something with better time complexity such as https://graphics.stanford.edu/~seander/bithacks.html#ReverseParallel

Reply:
Sure

Reply:
Hah I had noted this too, but didn't comment on it. I think I didn't even write it down for myself, which was dumb. Anyway...... the current implementation is fine in hardware, and a quick implementation would probably need a primitive. Unless you use `clashSimulation` from `Clash.Magic` to use the quick implementation in simulation but use `v2bv . reverse . bv2v` when generating HDL. That will write your primitive for you (but like a usual primitive, it's still up to you to make sure simulation and HDL match in behaviour).